### PR TITLE
Update config validation

### LINKED ISSUE #33690: Silent failure in generation parameters
### System Info

- `transformers` version: 4.44.2
- Platform: Linux-6.1.85+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.24.7
- Safetensors version: 0.4.5
- Accelerate version: 0.34.2
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.4.1+cu121 (False)
- Tensorflow version (GPU?): 2.17.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.8.5 (cpu)
- Jax version: 0.4.33
- JaxLib version: 0.4.33
- Using distributed or parallel set-up in script?: No

### Who can help?

@zucchini-nlp  @gante 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Hey! 

I noticed that `top_p` was silenctly failing so I tested the rest of the generation parameters and found that `no_repeat_ngram_size` also silently fails for the same reason:  the condition checks inside of the `_get_logits_processor()` method prevent their respective wrapper classes from executing, which is where the ValueError  are being raised.

For instance,  ``raise ValueError(f"`ngram_size` has to be a strictly positive integer, but is {ngram_size}")`` error is never reached when we set `no_repeat_ngram_size <= 0`.

Here is a simple example with the invalid values where generation proceeds without notifying the user. Ideally, those should raise errors or warnings.

```
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "EleutherAI/pythia-14m"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)
prompt = "hey there!"
inputs = tokenizer(prompt, return_tensors="pt")
generation_config = dict(do_sample=True, top_p=5, no_repeat_ngram_size=-1)

outputs = model.generate(input_ids=inputs['input_ids'], attention_mask=inputs['attention_mask'], **generation_config)

response = tokenizer.decode(outputs[0], skip_special_tokens=True)
response
```

### Expected behavior

- To not let things fail silently and proceed with a default value, instead raise a ValueError or issue a warning to the user.

- It would be great if the `generate` method could fail early when invalid values are passed, maybe by checking for them upfront in `_get_logits_processor`  before applying the generation parameters one by one and going through the entire process, this will help avoid wasting compute resources.

I would be happy to open a PR to help address this issue if that’s possible, thank you for all your work!

### PR DESCRIPTION
**Pull request by Manalelaidouni**
# What does this PR do?

Fixes #33690

This PR adds extra validation for generation parameters to fail early when invalid values are passed and not let things fail silently as suggested in the issue.  

## Before submitting

- [ ]  This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x]  Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
Pull Request section?
- [x]  Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
to it if that's the case.
- [ ]  Did you make sure to update the documentation with your changes? Here are the
[documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
[here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ]  Did you write any new necessary tests?

## Who can review?

@gante @zucchini-nlp

### CODE REVIEW

**Comment by Manalelaidouni** [comment]
The failing tests seem unrelated to my changes and all tests passed locally for me, is there something I'm missing here?

**Review by zucchini-nlp — COMMENTED** [review]
Cool, thanks for working on this! The tests failing are not related and re-trigerring run should help :)

**Comment by Manalelaidouni** [comment]
Hey @zucchini-nlp thanks! I’ve added extra validations for the remaining generation parameters.

I also validated `bad_words_ids` inside  `GenerationConfig.validate()` ,  to test it before running the forward pass, not at the end of it (when its `NoBadWordsLogitsProcessor` is executed).

For now, I’ve left the `_validate_arguments()` method in `NoBadWordsLogitsProcessor` intact, I was considering removing it since it’s now redundant, but I’d like get your green light before proceeding; If approved I would apply the same approach for `sequence_bias` and adjust their tests accordingly :)

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -574,17 +574,163 @@ def validate(self, is_init=False):
...
+            "max_length",
+            "min_length",
+            "no_repeat_ngram_size",
+            "encoder_no_repeat_ngram_size",
+        ]
+
+        probability_args = ["top_p", "typical_p", "min_p", "epsilon_cutoff", "eta_cutoff"]
+
+        float_pos_args = [
+            "temperature",
```
Comment (zucchini-nlp) [inline src/transformers/generation/configuration_utils.py:592]: temperature should be in probability args

**Review by zucchini-nlp — COMMENTED** [review]
Wow, an extensive validation list! Thanks!

>  I was considering removing it since it’s now redundant

No, I think we still need it because the logits processors can be initialized before `generate()` call and doesn't necessarily require a generation config.

Adding a few tests should be enough at this point :)

File: src/transformers/generation/configuration_utils.py (hunk 1 above)
Comment (Manalelaidouni) [reply src/transformers/generation/configuration_utils.py:592]: It seems that `temperature` shouldn’t be bound to a [0, 1] range because it scales down the logits (either dividing by a value greater than 1 or less than 1 to control randomness), but it doesn’t fit in `float_pos_args` either as `temperature`  can not be zero; it should be strictly float positive so it might need a separate test :)

Comment (zucchini-nlp) [reply src/transformers/generation/configuration_utils.py:592]: right, let's leave it as float positive argument then as the logit processor is added in all cases when `temp != 1`. In case users sets `0`, another error is raised when init the logit processor class

**Comment by Manalelaidouni** [comment]
Thanks @zucchini-nlp ! I believe all generation parameters are validated now

**Review by zucchini-nlp — APPROVED** [review]
@Manalelaidouni cool, LGTM! As said before adding a few tests should be enough at this point :)

You can add them in `tests/generation/test_configuration_utils.py` to make sure that incorrect variable through an error when passed to generation config. LMK if you need help with tests. Then feel free to tag the core maintainer @ ArthurZucker for final review

**Comment by Manalelaidouni** [comment]
Thank you @zucchini-nlp for reviewing, when adding tests I also made updates to some existing ones:

- So an older PR changed the `sequence_bias` format because it didn’t serialize properly, but the docstring and code still reference the dictionary format to users, which is confusing and since  `_convert_list_arguments_into_dict()` in `SequenceBiasLogitsProcessor`  is used for background compatibilty, the `_validate_arguments()` should raise a ` ValueError`  if the input is not in the list format, right?

- I also adjusted the `forced_bos_token_id` and `forced_eos_token_id` and added missing `get_generation_mode`, the rest of parameters I validated in the config seem be already tested.

Finally, I’m not entirely sure about how to address the `sequence_bias`  and   `bad_words_ids `  in the failing CI/CD tests, any guidance here would be appreciated, thank you again!

**Review by zucchini-nlp — COMMENTED** [review]
Thanks so much for iterating on this!

> since _convert_list_arguments_into_dict() in SequenceBiasLogitsProcessor is used for background compatibilty, the _validate_arguments() should raise a  ValueError if the input is not in the list format,

No, if we are keeping BC and the changes were done recently, we will not raise error if users pass in the old-style dict. It should be converted to the correct format to be used by logit processor at the end. But I see what you mean here, we won't be able to serialize the old dict format, so that means we'd need to `convert_to_list` in configuration here, which ensure that specific values are converted to serializable format in below code. Let's keep this for another PR, as the current one is getting big and should only check that correct values are passed

https://github.com/huggingface/transformers/blob/0b5b5e6a70249837293499e9363a64765a57111c/src/transformers/generation/configuration_utils.py#L1185 

> bad_words_ids  in the failing CI/CD tests

The token id in these args can be also a numpy int, see how we check the values here

https://github.com/huggingface/transformers/blob/0b5b5e6a70249837293499e9363a64765a57111c/src/transformers/generation/logits_process.py#L1307-L1310

**Comment by Manalelaidouni** [comment]
cc @ArthurZucker  whenever you have a moment, I'd appreciate your review, I wanted to see those CI tests green 😅 but it seems unrelated. Thanks!

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for the PR, this seems to be quite inflated 😓 
This is mostly not scalable 😢 so we should find another tool / effiecient way to define boundaries ands etc. Dataclasses might help!

**Comment by Manalelaidouni** [comment]
Thanks for reviewing this @ArthurZucker!

- dataclasses don't inherently validate data, it won’t raise an error if passed an invalid data type, so we would need to implement the validation logic manually anyway, relies on passing default model configs via kwargs which makes fully converting it to a dataclass impractical.

- Since you mentioned using a tool, a much simpler solution would to use `attrs` and it’s easily scalable, but it's adds an extra dependency ([Gante](https://github.com/huggingface/transformers/issues/33690#issuecomment-2444860878) advised against using `Pydantic` I assume because it significantly increases runtime).

- If we don't want to add dependencies, we can refactor the logic to a mapping approach which is scalable, something like this:

<br>

```python

class _ConfigValidator:

    VALIDATORS = {
        "pos_int_arg": (lambda x: isinstance(x, int) and x >= 0,
            "`{arg}` must be a positive integer, but got {value}."),

        "probability_arg": (lambda x: isinstance(x, (int, float)) and 0.0 <= x <= 1.0,
            "`{arg}` must be a float between 0 and 1, but got {value}."),
          ... }

    CONFIG_TYPES = {
        "top_k": "pos_int_arg",
        "top_p": "probability_arg",
         ...}


class GenerationConfig:
    ....

    def validate(self):
        for arg, value in self.__dict__.items():
            if arg in _ConfigValidator.CONFIG_TYPES:
                key = _ConfigValidator.CONFIG_TYPES[arg]
                validator, error_message = _ConfigValidator.VALIDATORS[key]
                if value is not None and not validator(value):
                    raise ValueError(error_message.format(arg=arg, value=value))


```

<br>

`_ConfigValidator` would hold also more complex validation for consistency and keep `GenerationConfig.validate()` less inflated for any future generation arguments. Let me know what you think before I move forward :)

**Comment by ArthurZucker** [comment]
I really don't mind using pydantic, it's become quite important and we have so many configs that we need to validate! Do you mind drafting a PR for that? 🤗

**Comment by Manalelaidouni** [comment]
Hey @ArthurZucker, apologies for the delayed response, I just opened a PR for it!

**Comment by gante** [comment]
@Manalelaidouni this PR is orthogonal to the pydantic PR, correct? Or should we merge one before the other?

**Comment by Manalelaidouni** [comment]
@gante they're indeed orthogonal so I'll close this PR :)
